src/compute-client/src/metrics.rs

```diff
@@ -0,0 +1,126 @@
+}
+
+impl ComputeControllerMetrics {
+    pub fn new(metrics_registry: MetricsRegistry) -> Self {
+        ComputeControllerMetrics {
+            messages_sent_bytes: metrics_registry.register(metric!(
+                name: "mz_compute_messages_sent_bytes",
+                help: "size of compute messages sent",
+                var_labels: ["instance", "replica"]
+            )),
+            messages_received_bytes: metrics_registry.register(metric!(
+                name: "mz_compute_messages_received_bytes",
+                help: "size of compute messages received",
+                var_labels: ["instance", "replica"]
+            )),
+        }
+    }
+
+    pub fn for_instance(&self, instance_id: ComputeInstanceId) -> InstanceMetrics {
+        InstanceMetrics {
+            messages_sent_bytes: self.messages_sent_bytes.clone(),
+            messages_received_bytes: self.messages_received_bytes.clone(),
+            instance_id,
+        }
+    }
+}
+
+/// Per-instance metrics
+#[derive(Debug, Clone)]
+pub struct InstanceMetrics {
+    messages_sent_bytes: prometheus::HistogramVec,
+    messages_received_bytes: prometheus::HistogramVec,
+    instance_id: ComputeInstanceId,
+}
+
+impl InstanceMetrics {
+    pub fn for_replica(&self, replica_id: ReplicaId) -> ReplicaMetrics {
+        let labels = vec![self.instance_id.to_string(), replica_id.to_string()];
+        let messages_sent_bytes = self
+            .messages_sent_bytes
+            .get_delete_on_drop_histogram(labels.clone());
+        let messages_received_bytes = self
+            .messages_received_bytes
+            .get_delete_on_drop_histogram(labels);
+
+        ReplicaMetrics {
+            inner: Arc::new(ReplicaMetricsInner {
+                messages_sent_bytes,
+                messages_received_bytes,
+            }),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct ReplicaMetricsInner {
+    messages_sent_bytes: DeleteOnDropHistogram<'static, Vec<String>>,
+    messages_received_bytes: DeleteOnDropHistogram<'static, Vec<String>>,
+}
+
+/// Per-replica metrics.
+#[derive(Debug, Clone)]
+pub struct ReplicaMetrics {
+    inner: Arc<ReplicaMetricsInner>,
+}
+
+impl ReplicaMetrics {
+    pub fn for_connection(&self, _index: usize) -> ReplicaConnectionMetrics {
+        ReplicaConnectionMetrics {
+            inner: Arc::clone(&self.inner),
+        }
+    }
+}
+
+/// Per-replica metrics
+#[derive(Clone, Debug)]
+pub struct ReplicaConnectionMetrics {
+    inner: Arc<ReplicaMetricsInner>,
+}
+
+/// Make ReplicaConnectionMetric pluggable into the gRPC connection.
+impl StatsCollector<ProtoComputeCommand, ProtoComputeResponse> for ReplicaConnectionMetrics {
+    fn send_event(&self, _item: &ProtoComputeCommand, size: usize) {
+        match f64::try_cast_from(u64::cast_from(size)) {
+            Some(x) => self.inner.messages_sent_bytes.observe(x),
+            None => tracing::warn!("size outside range of f64, ignoring message"),
```

Comment:
This warning is a bit confusing, since the range (`MIN..=MAX`) of `f64` is way larger than that of `u64`. Maybe we could clarify it to say that the integer value is not representable by `f64`?

Also, I think it would be a good idea to include the offending value in the log message.